dali/dfu/dfuutil.cpp

```diff
@@ -857,21 +857,61 @@ class CDFUhelper: public CInterface, implements IDFUhelper
             else
                 throwError1(DFUERR_DFileNotFound, lfn);
         }
-        else {
-            Owned<IFileDescriptor> fdesc = file->getFileDescriptor();
-            Owned<IPropertyTree> t = fdesc->getFileTree();
+        else
+        {
+            Owned<IPropertyTree> t = queryDistributedFileDirectory().getFileTree(lfn, user);
             toXML(t, out, true);
         }
         return out;
     }
 
-    void addFileXML(const char *lfn,const StringBuffer &xml,IUserDescriptor *user)
+    void addFileXML(const char *lfn, const StringBuffer &xml, IUserDescriptor *user)
     {
         Owned<IPropertyTree> t = createPTreeFromXMLString(xml);
-        Owned<IFileDescriptor> fdesc = deserializeFileDescriptorTree(t,&queryNamedGroupStore(),0);
-        Owned<IDistributedFile> file = queryDistributedFileDirectory().createNew(fdesc,true);
-        if (file)
-            file->attach(lfn,user);
+        IDistributedFileDirectory &dfd = queryDistributedFileDirectory();
+        // Check if this XML is a superfile map
+        if (t->hasProp("@numsubfiles"))
```

Comment:
It would be cleaner if did:
```
        DfsXmlBranchKind kind;
        if (0 == strcmp(t->queryName(), queryDfsXmlBranchName(DXB_SuperFile))
            kind = DXB_SuperFile;
        else if (0 == strcmp(t->queryName(), queryDfsXmlBranchName(DXB_File))
            kind = DXB_File;
        else
            throw MakeStringException(-1, "Unrecognised file XML root tag detected: '%s'", t->queryName());
```

Then use kind conditionally above where needed.